File: 2022/day08/javascript/part2.mjs
import { join } from "node:path";
import { readFile } from "node:fs/promises";

const filePath = new URL(join(import.meta.url, "..", "..", "input.txt"));

const file = await readFile(filePath, { encoding: "utf-8" });
const data = file.split("\n");

const treeMap = data.map((row) => row.split("").map(Number));
const scoreList = [];

for (let i = 1; i < data.length - 1; i++) {
  for (let j = 1; j < data[i].length - 1; j++) {
    const tree = treeMap[i][j];

    const scores = [0, 0, 0, 0];

    for (let k = i - 1; k > -1; k--) {
      scores[0] += 1;
      if (treeMap[k][j] >= tree) break;
    }

    for (let k = i + 1; k < treeMap.length; k++) {
      scores[1] += 1;
      if (treeMap[k][j] >= tree) break;
    }

    for (const k of [...treeMap[i]].splice(0, j).reverse()) {
      scores[2] += 1;
      if (k >= tree) break;
    }

    for (const k of [...treeMap[i]].splice(j + 1, treeMap[i].length)) {
      scores[3] += 1;
      if (k >= tree) break;
    }

    scoreList.push(scores.reduce((a, i) => a * i, 1));
  }
}

console.log(Math.max(...scoreList));

File: 2022/day08/python/part2.py
from pathlib import Path
from math import prod
file_path = Path(__file__, '..', '..', 'input.txt').resolve()


with open(file_path) as f:
    data = f.read().splitlines()

    score_list = []
    tree_map = [list(map(int, x)) for x in data]

    for i in range(1, len(tree_map) - 1):
        for j in range(1, len(tree_map[i]) - 1):
            tree = tree_map[i][j]
            scores = [0, 0, 0, 0]

            for k in range(i - 1, -1, -1):
                scores[0] += 1
                if tree_map[k][j] >= tree:
                    break

            for k in range(i + 1, len(tree_map)):
                scores[1] += 1
                if tree_map[k][j] >= tree:
                    break

            for k in tree_map[i][:j][::-1]:
                scores[2] += 1
                if k >= tree:
                    break

            for k in tree_map[i][j + 1:]:
                scores[3] += 1
                if k >= tree:
                    break

            score_list.append(prod(scores))

    print(max(score_list))
